### PR TITLE
Multipart messaging

### PR DESCRIPTION
**Pull request by rudigiesler**
This adds the multiple strategies for splitting out long messages into multiple parts.

For the error handling, it does it a bit differently than Vumi. In vumi, we send all the parts, not waiting for the ROK response for each. And then as those come in, we wait for all the responses, and then send an ack or nack.

Here, we send the parts one at a time, waiting for the ROK for each part. And as soon as we receive an error repsonse, we send a nack and stop sending the parts.

This seems like a better approach, and maybe it wasn't done in vumi because of the different interface for the send_pdu function, but I'm not sure if there are any downsides to doing it this way instead.

We're also splitting the messages at different points. In vumi, it was a hardcoded 130 chars. Here, we split according to the encoding type.

Another difference is that if it fits in one message, then we send it as a normal message. This will mean that if we select a multipart strategy, we won't split if it doesn't fit in one part of a multipart message, but does fit in a single message, which seems like the behaviour that we would want.

### CODE REVIEW

File: src/vumi2/transports/smpp/client.py [hunk 1]
```diff
@@ -211,28 +211,29 @@ async def send_pdu(
...
-            status_code = command_status_name_map[response.status.name]
-            nack_reason = command_status_value_map[status_code]["description"]
-            event = Event(
-                user_message_id=message.message_id,
-                sent_message_id=message.message_id,
-                event_type=EventType.NACK,
-                nack_reason=nack_reason,
-            )
+            # We will always get a response from a request
+            response = cast(PDU, response)
```
Comment (jerith) [inline src/vumi2/transports/smpp/client.py:220]: Is there value in casting to something more specific than this?

File: src/vumi2/transports/smpp/processors.py [hunk 2]
```diff
@@ -87,32 +87,175 @@ def __init__(self, config: dict, sequencer: Sequencer) -> None:
...
+    async def submit_sm_short_message(
+        self, from_addr: str, to_addr: str, short_message: bytes
+    ) -> SubmitSM:
+        return SubmitSM(
+            seqNum=await self.sequencer.get_next_sequence_number(),
+            service_type=self.config.service_type,
+            source_addr_ton=self.config.source_addr_ton,
+            source_addr_npi=self.config.source_addr_npi,
+            source_addr=from_addr,
+            dest_addr_ton=self.config.dest_addr_ton,
```
Comment (jerith) [inline src/vumi2/transports/smpp/processors.py:150]: Is there some commonality we can extract here? Maybe an `async def _submit_sm(self, ...)` method that handles all the statically configured fields?

File: tests/transports/smpp/test_processors.py [hunk 3]
```diff
@@ -109,3 +112,90 @@ async def test_submit_sm_outbound_vumi_message_custom_config(
...
+    assert pdu2.params["sar_total_segments"] == 2
+    assert pdu2.params["sar_segment_seqnum"] == 2
+
+
+async def test_submit_sm_outbound_vumi_message_udh(
+    submit_sm_processor: SubmitShortMessageProcessor,
+):
+    """
+    Creates a valid PDU representing the outbound vumi message, storing the contents of
+    long messages into multiple PDUs with sar parameters
```
Comment (jerith) [inline tests/transports/smpp/test_processors.py:176]: `udh parameters`?

**Review by jerith — COMMENTED** [review]
Aside from my inline comments, I think we also need tests for various sar/udh edge cases. (Messages short enough to fit in a single PDU, different encodings, etc.)

File: src/vumi2/transports/smpp/client.py (hunk 1 above)
Comment (rudigiesler) [reply src/vumi2/transports/smpp/client.py:220]: Maybe for autocomplete? Although the library isn't really great with that, even if you choose a specific PDU type, it's just the same base PDU fields, and then everything is in a dict-like params, so you don't get much. But we could do a `PDUResponse`, or since we know that we're always doing a `submit_sm`, maybe a `SubmitSMResp`, but since the code that creates the PDUs is separate to this, I'm worried that if in future it gets changed to give other kinds of PDUs, that we forget to change this.

What I really want is for `send_pdu` to have a dynamic type. If it's a request, then it returns a response. If it's a response, then it returns None. I tried to do this with type overloads, but it gives an error that the second overload will never match, and I can't understand why.

I think maybe I'll change that to a response, instead of the base PDU

File: src/vumi2/transports/smpp/processors.py (hunk 2 above)
Comment (rudigiesler) [reply src/vumi2/transports/smpp/processors.py:150]: Yeah I think there is, let me take a crack at that

**Review by jerith — APPROVED** [review]
:+1:
